jobs/io/stub_io.py

```diff
@@ -19,52 +19,289 @@
+    there isn't any state that can be shared between them.
+
+    This restriction gives us the added benefit of being able to enforce tighter
+    restrictions on how datastore operations are executed (since the real module
+    behaves in the same way; i.e. communicating via RPC), specifically:
+        1.  Reads should only occur at the very beginning of a Pipeline.
+        2.  Writes/Deletes should only occur at the very end of a Pipeline.
+        3.  Reads should never be performed after a Write or Delete (because
+            the ordering of PTransform operations are not guaranteed).
+    """
 
     def __init__(self):
-        self._models_by_id = {}
+        self._server = python_utils.SimpleXMLRPCServer(
+            # Using 0 instructs the OS to acquire a free port on our behalf.
+            ('localhost', 0),
+            # We need to allow None because that is a valid NDB Property value.
+            allow_none=True)
+        self._server.register_function(
+            self._read_from_datastore_handler, name='ReadFromDatastore')
+        self._server.register_function(
+            self._write_to_datastore_handler, name='WriteToDatastore')
+        self._server.register_function(
+            self._delete_from_datastore_handler, name='DeleteFromDatastore')
+
+        self._server_port = self._server.server_address[1]
+        self._server_context_is_acquired = False
+        self._models = {}
+        self._models_lock = threading.Lock()
+        self._write_was_called = False
+        self._delete_was_called = False
+
+    @contextlib.contextmanager
+    def context(self):
+        """Returns a context in which the stub listens to XML-RPC requests."""
+        self._server_context_is_acquired = True
+        thread = threading.Thread(target=self._server.serve_forever)
+        thread.start()
+        try:
+            yield
+        finally:
+            self._server.shutdown()
+            thread.join()
+            self._server_context_is_acquired = False
+
+    def get_everything(self):
+        """Returns all models currently stored in the stub.
+
+        Returns:
+            list(Model). All of the models in the stub.
+        """
+        with self._models_lock:
+            return list(self._models.values())
+
+    def put_multi(self, models):
+        """Puts the input models into the stub.
+
+        Args:
+            models: list(Model). The NDB models to put into the stub.
+        """
+        with self._models_lock:
+            self._models.update({model.key: model for model in models})
+
+    def delete_multi(self, models):
+        """Deletes the models from the stub, if they exist.
+
+        Args:
+            models: list(Model). The models to delete from the stub.
+        """
+        with self._models_lock:
+            for model in models:
+                self._models.pop(model.key, None)
+
+    def ReadFromDatastore(self): # pylint: disable=invalid-name
```

Comment:
Why do these and the next 2 functions not follow naming conventions?

(In general, whenever you add a pylint-disable, I suggest writing a review comment to explain to reviewers why it's needed -- or possibly a code comment if you think future devs might have the same question.)

Reply:
Done, have added a code comment.